app/Actions/Album/Delete.php

```diff
@@ -2,29 +2,130 @@
-use App\Models\Extensions\BaseAlbum;
+use App\Actions\Photo\Delete as PhotoDelete;
+use App\Facades\AccessControl;
+use App\Image\FileDeleter;
+use App\Models\Album;
+use App\Models\BaseAlbumImpl;
+use App\Models\TagAlbum;
 use App\SmartAlbums\UnsortedAlbum;
+use Illuminate\Database\Query\Builder as BaseBuilder;
 
+/**
+ * Deletes the albums with the designated IDs **efficiently**.
+ *
+ * This class deliberately violates the principle of separations of concerns.
+ * In an ideal world, the method would simply call `->delete()` on every
+ * `Album` model, the `Album` model would take care of deleting its
+ * sub-albums and every album in turn would take care of deleting its photos.
+ * But this is extremely inefficient due to Laravel's architecture:
+ *
+ *  - Models are heavyweight god classes such that every instance also carries
+ *    the whole code for serialization/deserialization
+ *  - Models are active records (and don't use the unit-of-work pattern), i.e.
+ *    every deletion of a model directly triggers a DB operation; they are
+ *    not deferred into a batch operation
+ *
+ * Moreover, while removing the records for albums and photos from the
+ * DB can be implemented rather efficiently, the actual file operations may
+ * take some time.
+ * Especially, if the files are not stored locally but on a remote file system.
+ * Hence, this method collects all files which need to be removed.
+ * The caller can then decide to delete them asynchronously.
+ */
 class Delete extends Action
 {
 	/**
-	 * @param array $albumIDs
+	 * Deletes the designated albums from the DB.
 	 *
-	 * @return bool
+	 * The method only deletes the records for albums, photos, their size
+	 * variants and potentially associated symbolic links from the DB.
+	 * The method does not delete the associated files from the physical
+	 * storage.
+	 * Instead, the method returns an object in which all these files have
+	 * been collected.
+	 * This object can (and must) be used to eventually delete the files,
+	 * however doing so can be deferred.
+	 *
+	 * @param string[] $albumIDs the album IDs
+	 *
+	 * @return FileDeleter contains the collected files which became obsolete
 	 */
-	public function do(array $albumIDs): bool
+	public function do(array $albumIDs): FileDeleter
 	{
-		$albums = $this->albumFactory->findWhereIDsIn($albumIDs);
-		$success = true;
+		try {
+			$photoIDs = [];
 
-		/** @var AbstractAlbum $album */
-		foreach ($albums as $album) {
-			if ($album instanceof BaseAlbum || $album instanceof UnsortedAlbum) {
-				$success &= $album->delete();
+			// Among the smart albums, the unsorted album is special,
+			// because it provides deletion of photos
+			if (in_array(UnsortedAlbum::ID, $albumIDs, true)) {
+				$query = UnsortedAlbum::getInstance()->photos();
+				if (!AccessControl::is_admin()) {
+					$query->where('owner_id', '=', AccessControl::id());
+				}
+				$photoIDs = UnsortedAlbum::getInstance()->photos()->pluck('id')->all();
```

Comment:
```suggestion
				$photoIDs = $query->pluck('id')->all();
```

Untested, but we currently aren't using `$query` anywhere?!